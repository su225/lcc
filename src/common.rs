#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Location {
    pub(crate) line: usize,
    pub(crate) column: usize,
}

impl Location {
    pub(crate) fn advance_line(&mut self) {
        self.line += 1;
        self.column = 1;
    }

    pub(crate) fn advance_tab(&mut self) {
        self.column = ((self.column + 7) / 8) * 8;
    }

    pub(crate) fn advance(&mut self) {
        self.column += 1;
    }
}

#[derive(Debug, Eq, PartialEq, Hash, Clone, Copy)]
pub enum Radix {
    Binary,
    Octal,
    Decimal,
    Hexadecimal,
}