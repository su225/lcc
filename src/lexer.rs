use std::collections::HashMap;
use std::fmt::{Display, Formatter};
use std::iter::Peekable;
use std::str::Chars;

use once_cell::sync::Lazy;
use thiserror::Error;

#[derive(Debug, Eq, PartialEq, Clone, Hash)]
pub enum KeywordIdentifier {
    TypeInt,
    TypeVoid,
    Return,
}

type Radix = u8;

const DECIMAL: Radix = 10;
const OCTAL: Radix = 8;
const HEXADECIMAL: Radix = 16;
const BINARY: Radix = 2;

#[derive(Debug, Eq, PartialEq)]
pub enum TokenType<'a> {
    Keyword(KeywordIdentifier),
    OpenParentheses,
    CloseParentheses,
    OpenBrace,
    CloseBrace,
    Semicolon,
    Identifier(&'a str),
    IntConstant(&'a str, Radix),
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
struct Location {
    line: usize,
    column: usize,
}

impl Location {
    fn advance_line(&mut self) {
        self.line += 1;
        self.column = 1;
    }

    fn advance_tab(&mut self) {
        self.column = ((self.column + 7) / 8) * 8;
    }

    fn advance(&mut self) {
        self.column += 1;
    }
}

#[derive(Error, Debug, PartialEq)]
pub enum LexerError {
    #[error("{location:?}: unexpected character")]
    UnexpectedCharacter { location: Location },

    #[error("{location:?}: unknown radix representation of integer")]
    UnknownRadixRepresentation { location: Location },

    #[error("{location:?}: invalid digit {digit:?} for radix {radix:?}")]
    InvalidDigitForRadix { location: Location, digit: char, radix: Radix },

    #[error("{location:?}: invalid character {ch:?} for identifier")]
    InvalidIdentifierCharacter { location: Location, ch: char },
}

type LexerResult<T> = Result<T, LexerError>;

static KEYWORDS: Lazy<HashMap<&'static str, KeywordIdentifier>> = Lazy::new(|| {
    HashMap::from([
        ("int", KeywordIdentifier::TypeInt),
        ("void", KeywordIdentifier::TypeVoid),

        ("return", KeywordIdentifier::Return),
    ])
});

static KEYWORD_STRINGS: Lazy<HashMap<KeywordIdentifier, &'static str>> = Lazy::new(|| {
    HashMap::from([
        (KeywordIdentifier::TypeInt, "int"),
        (KeywordIdentifier::TypeVoid, "void"),

        (KeywordIdentifier::Return, "return"),
    ])
});

impl<'a> Display for TokenType<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match &self {
            TokenType::Keyword(k) => f.write_str(KEYWORD_STRINGS.get(k).unwrap()),
            TokenType::OpenParentheses => f.write_str("("),
            TokenType::CloseParentheses => f.write_str(")"),
            TokenType::OpenBrace => f.write_str("{"),
            TokenType::CloseBrace => f.write_str("}"),
            TokenType::Semicolon => f.write_str(";"),
            TokenType::Identifier(x) => f.write_fmt(format_args!("identifier:{}", x)),
            TokenType::IntConstant(x, radix) => f.write_fmt(format_args!("int:[{}, radix:{}]", x, radix)),
        }
    }
}

#[derive(Debug, Eq, PartialEq)]
pub struct Token<'a> {
    token_type: TokenType<'a>,
    location: Location,
}

pub struct Lexer<'a> {
    input: &'a str,
    char_stream: Peekable<Chars<'a>>,
    cur_stream_pos: usize,
    cur_location: Location,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            input,
            char_stream: input.chars().peekable(),
            cur_stream_pos: 0,
            cur_location: Location { line: 1, column: 1 },
        }
    }

    fn next_char(&mut self) -> Option<char> {
        let ch = self.char_stream.next();
        if ch.is_none() {
            return None;
        }
        let c = ch.unwrap();
        self.cur_stream_pos += c.len_utf8();
        match c {
            '\n' => self.cur_location.advance_line(),
            '\t' => self.cur_location.advance_tab(),
            _ => self.cur_location.advance(),
        }
        return ch;
    }

    fn tokenize_single_char(&mut self, token_type: TokenType<'a>) -> Token<'a>  {
        let token = Token {
            token_type,
            location: self.cur_location.clone(),
        };
        self.next_char();
        token
    }

    fn tokenize_integer_constant(&mut self) -> Result<Token<'a>, LexerError> {
        let start_loc = self.cur_location.clone();
        let start_pos = self.cur_stream_pos;
        let first_digit = self.next_char().unwrap();
        let mut expected_radix = OCTAL;
        if first_digit != '0' {
            expected_radix = DECIMAL;
        } else {
            let second_digit = self.next_char();
            if second_digit.is_none() {
                return Ok(Token { location: start_loc, token_type: TokenType::IntConstant("0", DECIMAL) });
            }
            let digit2 = second_digit.unwrap();
            if digit2 == 'x' || digit2 == 'X' {
                expected_radix = HEXADECIMAL;
            } else if digit2 == 'b' || digit2 == 'B' {
                expected_radix = BINARY;
            } else if !digit2.is_digit(OCTAL as u32) {
                // If it is neither 0x, 0X, 0b, 0B or 0[Octal digit]
                // then it is an invalid number. Hence, it is an error.
                return Ok(Token { location: start_loc, token_type: TokenType::IntConstant("0", DECIMAL) });
            }
        }
        // Once we have determined the radix, we iterate through the digits as long
        // as we find a valid stop point: semicolon, whitespace (space, new-line, tab), end.
        // If we stop at characters that are not valid digits, then we throw an error. We
        // don't support digit grouping yet to keep the lexer simple.
        loop {
            let loc = self.cur_location.clone();
            let next = self.char_stream.peek();
            if next.is_none() {
                break;
            }
            let n = next.unwrap().clone();
            if !n.is_digit(expected_radix as u32) {
                break;
            }
            self.next_char();
        }
        return Ok(Token {
            location: start_loc,
            token_type: TokenType::IntConstant(&self.input[start_pos..self.cur_stream_pos], expected_radix),
        })
    }

    fn tokenize_identifier_or_keyword(&mut self) -> Result<Token<'a>, LexerError> {
        let start_loc = self.cur_location;
        let start_pos = self.cur_stream_pos;
        let first_char = self.next_char().unwrap();
        if first_char != '_' && !first_char.is_alphabetic() {
            return Err(LexerError::InvalidIdentifierCharacter{ location: start_loc, ch: first_char });
        }
        loop {
            let cur_loc = self.cur_location;
            let ch = self.char_stream.peek();
            if ch.is_none() {
                break;
            }
            let c = ch.unwrap().clone();
            if !c.is_alphanumeric() && c != '_' {
                break;
            }
            self.next_char();
        }
        let word = &self.input[start_pos..self.cur_stream_pos];
        match KEYWORDS.get(word) {
            None => Ok(Token{
                location: start_loc,
                token_type: TokenType::Identifier(word),
            }),
            Some(k) => Ok(Token{
                location: start_loc,
                token_type: TokenType::Keyword(k.clone()),
            })
        }
    }

    fn next_token(&mut self) -> Result<Option<Token<'a>>, LexerError> {
        loop {
            let cur = self.char_stream.peek();
            if cur.is_none() {
                // We have reached the end of the stream. Hence, we cannot
                // tokenize anymore. So we just return None
                return Ok(None);
            }
            let cur_char = cur.unwrap();
            match *cur_char {
                ';' => {
                    let token = self.tokenize_single_char(TokenType::Semicolon);
                    return Ok(Some(token));
                },
                '(' => {
                    let token = self.tokenize_single_char(TokenType::OpenParentheses);
                    return Ok(Some(token));
                },
                ')' => {
                    let token = self.tokenize_single_char(TokenType::CloseParentheses);
                    return Ok(Some(token));
                },
                '{' => {
                    let token = self.tokenize_single_char(TokenType::OpenBrace);
                    return Ok(Some(token));
                },
                '}' => {
                    let token = self.tokenize_single_char(TokenType::CloseBrace);
                    return Ok(Some(token));
                },
                '0'..='9' => {
                    let token = self.tokenize_integer_constant()?;
                    return Ok(Some(token));
                },
                '\n' | '\t' | ' ' => {
                    self.next_char();
                    continue;
                },
                _ => {
                    let token = self.tokenize_identifier_or_keyword()?;
                    return Ok(Some(token));
                },
            }
        }
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = Result<Token<'a>, LexerError>;

    fn next(&mut self) -> Option<Self::Item> {
        let next_tok = self.next_token();
        match next_tok {
            Ok(Some(tok)) => Some(Ok(tok)),
            Ok(None) => None,
            Err(e) => Some(Err(e)),
        }
    }
}

#[cfg(test)]
mod test {
    use std::collections::HashMap;

    use crate::lexer::{BINARY, DECIMAL, HEXADECIMAL, KEYWORDS, Lexer, LexerError, LexerResult, Location, OCTAL, Token, TokenType};
    use crate::lexer::TokenType::Keyword;

    #[test]
    fn test_tokenizing_open_and_close_parentheses() {
        let source = "()";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenParentheses, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseParentheses, location: Location { line: 1, column: 2 }},
        ]));
    }

    #[test]
    fn test_tokenizing_open_and_close_braces() {
        let source = "{}";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenBrace, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseBrace, location: Location { line: 1, column: 2 }},
        ]));
    }

    #[test]
    fn test_tokenizing_with_newlines() {
        let source = "(\n\n)";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenParentheses, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseParentheses, location: Location { line: 3, column: 1 }},
        ]));
    }

    #[test]
    fn test_tokenizing_with_tabs() {
        let source = "(\t)";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenParentheses, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseParentheses, location: Location { line: 1, column: 8 }},
        ]));
    }

    #[test]
    fn test_tokenizing_with_whitespaces() {
        let source = "(   )";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::OpenParentheses, location: Location { line: 1, column: 1 }},
            Token { token_type: TokenType::CloseParentheses, location: Location { line: 1, column: 5 }},
        ]));
    }

    #[test]
    fn test_tokenizing_semicolon() {
        let source = ";";
        let lexer = Lexer::new(source);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        assert_eq!(tokens, Ok(vec![
            Token { token_type: TokenType::Semicolon, location: Location { line: 1, column: 1 }},
        ]));
    }

    #[test]
    fn test_tokenizing_valid_identifiers() {
        let identifiers = vec![
            "abcde",
            "abcde123",
            "hello_world_123",
            "helloWorld123",
            "_abcde",
            "_123",
            "_123_456",
            "café",
            "αριθμός",
            "число",
            "数字",
        ];
        for src in identifiers.into_iter() {
            let lexer = Lexer::new(src);
            let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
            let expected_tokens: LexerResult<Vec<Token>> = Ok(vec![
                Token { token_type: TokenType::Identifier(src), location: Location { line: 1, column: 1 }},
            ]);
            assert_eq!(tokens, expected_tokens,
                       "lexing identifier {}: expected: {:?}, actual:{:?}",
                        src, expected_tokens, tokens);
        }
    }

    #[test]
    fn test_tokenizing_keywords() {
        for (kw, kwid) in KEYWORDS.iter() {
            let lexer = Lexer::new(kw);
            let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
            let expected_tokens: LexerResult<Vec<Token>> = Ok(vec![
                Token { token_type: Keyword(kwid.clone()), location: Location { line: 1, column: 1 } },
            ]);
            assert_eq!(tokens, expected_tokens,
                      "lexing keyword identifier {}: expected: {:?}, actual:{:?}",
                      kw, expected_tokens, tokens);
        }
    }

    #[test]
    fn test_tokenizing_valid_integers() {
        let integers_base10 = vec![
            "1",
            "0",
            "2",
            "3",
            "44",
            "55",
            "66",
            "777",
            "888",
            "9189",
            "189087931798698368761873",
            "0\t\t",
            "0  ",
            "0\n",
        ];

        let integers_hex = vec!["0xdeadbeef",
            "0Xcafebabe",
            "0XDEADBEEF",
            "0xCAFEbabe",
        ];

        let integers_octal = vec![
            "000",
            "01234567",
            "012300",
        ];

        let integers_binary = vec![
            "0b01010010010",
            "0b001001001",
        ];

        let int_tests = HashMap::from([
            (10, integers_base10),
            (16, integers_hex),
            (8, integers_octal),
            (2, integers_binary),
        ]);


        for (base, srcs) in int_tests.into_iter() {
            for src in srcs.into_iter() {
                let lexer = Lexer::new(src);
                let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
                let expected_tokens = Ok(vec![
                    Token { token_type: TokenType::IntConstant(src.trim(), base), location: Location { line: 1, column: 1 }},
                ]);
                assert_eq!(tokens, expected_tokens,
                           "lexing identifier {}: expected: {:?}, actual:{:?}",
                           src, expected_tokens, tokens);
            }
        }
    }

    #[test]
    fn test_bad_identifier_error() {
        let src = "abcd@";
        let lexer = Lexer::new(src);
        let tokens: LexerResult<Vec<Token>> = lexer.into_iter().collect();
        let expected = Err(LexerError::InvalidIdentifierCharacter {
            location: Location {line: 1, column: 5},
            ch: '@',
        });
        assert_eq!(tokens, expected);
    }
}