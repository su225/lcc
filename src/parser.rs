//! Parser for the language: This implements parser for a subset of
//! features in C programming language. A simple Recursive Descent
//! Parsing is used. It is handwritten.

use std::iter::Peekable;

use thiserror::Error;

use crate::common::{Location, Radix};
use crate::lexer::{KeywordIdentifier, Lexer, LexerError, Token, TokenTag, TokenType};

#[derive(Debug, PartialEq)]
pub struct Symbol<'a> {
    pub name: &'a str,
    location: Location,
}

#[derive(Debug, PartialEq)]
pub(crate) enum UnaryOperator {
    Complement,
    Negate,
}

#[derive(Debug, PartialEq)]
pub(crate) enum BinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
}

#[derive(Debug, PartialEq)]
pub(crate) enum ExpressionKind<'a> {
    IntConstant(&'a str, Radix),
    Unary(UnaryOperator, Box<Expression<'a>>),
}

#[derive(Debug, PartialEq)]
pub(crate) struct Expression<'a> {
    location: Location,
    pub(crate) kind: ExpressionKind<'a>,
}

#[derive(Debug, PartialEq)]
pub(crate) enum PrimitiveKind {
    Integer,
    UnsignedInteger,
    LongInteger,
}

#[derive(Debug, PartialEq)]
pub(crate) enum TypeExpressionKind {
    Primitive(PrimitiveKind),
}

#[derive(Debug, PartialEq)]
pub(crate) struct TypeExpression {
    pub(crate) location: Location,
    pub(crate) kind: TypeExpressionKind,
}

#[derive(Debug, PartialEq)]
pub(crate) enum StatementKind<'a> {
    Return(Expression<'a>),
}

#[derive(Debug, PartialEq)]
pub(crate) struct Statement<'a> {
    pub(crate) location: Location,
    pub(crate) kind: StatementKind<'a>,
}

#[derive(Debug, PartialEq)]
pub(crate) struct FunctionDefinition<'a> {
    pub(crate) location: Location,
    pub(crate) name: Symbol<'a>,
    pub(crate) body: Vec<Statement<'a>>,
}

#[derive(Debug, PartialEq)]
pub(crate) struct ProgramDefinition<'a> {
    pub(crate) functions: Vec<FunctionDefinition<'a>>,
}

pub struct Parser<'a> {
    token_provider: Peekable<Lexer<'a>>,
}

#[derive(Error, Debug, PartialEq)]
pub enum ParserError {
    #[error("error from lexer: {0}")]
    TokenizationError(LexerError),

    #[error("keyword {kwd:?} used as identifier")]
    KeywordUsedAsIdentifier { location: Location, kwd: KeywordIdentifier },

    #[error("{location:?}: unexpected token")]
    UnexpectedToken { location: Location, expected_token_tags: Vec<TokenTag> },

    #[error("unexpected end of file. Expected {:?}", .0)]
    UnexpectedEnd(Vec<TokenTag>),

    #[error("{location:?}: expected keyword {keyword_identifier:?}")]
    ExpectedKeyword {
        location: Location,
        keyword_identifier: KeywordIdentifier,
        actual_token: TokenTag,
    },
}

impl<'a> Parser<'a> {
    pub fn new(token_provider: Lexer<'a>) -> Parser<'a> {
        Parser { token_provider: token_provider.peekable() }
    }

    /// parse parses the given source file and returns the
    /// Abstract Syntax Tree (AST).
    pub fn parse(&mut self) -> Result<ProgramDefinition<'a>, ParserError> {
        Ok(self.parse_program()?)
    }

    fn parse_program(&mut self) -> Result<ProgramDefinition<'a>, ParserError> {
        let mut functions: Vec<FunctionDefinition<'a>> = vec![];
        loop {
            if self.token_provider.peek().is_none() {
                break; // No more tokens and hence we are safe to stop here
            }
            // Otherwise parse the next function definition
            let func_definition = self.parse_function_definition()?;
            functions.push(func_definition);
        }

        Ok(ProgramDefinition { functions })
    }

    fn parse_function_definition(&mut self) -> Result<FunctionDefinition<'a>, ParserError> {
        let return_type = self.parse_type_expression()?;
        let name = self.parse_identifier()?;
        self.expect_open_parentheses()?;
        self.parse_function_parameters()?;
        self.expect_close_parentheses()?;
        let body = self.parse_function_body()?;
        Ok(FunctionDefinition { location: return_type.location, name, body })
    }

    fn parse_function_parameters(&mut self) -> Result<(), ParserError> {
        self.expect_keyword(KeywordIdentifier::TypeVoid)?;
        Ok(())
    }

    fn parse_function_body(&mut self) -> Result<Vec<Statement<'a>>, ParserError> {
        self.expect_open_braces()?;
        let stmt = self.parse_statement()?;
        self.expect_close_braces()?;
        Ok(vec![stmt])
    }

    fn expect_open_parentheses(&mut self) -> Result<(), ParserError> {
        self.expect_token_with_tag(TokenTag::OpenParentheses)
    }

    fn expect_close_parentheses(&mut self) -> Result<(), ParserError> {
        self.expect_token_with_tag(TokenTag::CloseParentheses)
    }

    fn expect_open_braces(&mut self) -> Result<(), ParserError> {
        self.expect_token_with_tag(TokenTag::OpenBrace)
    }

    fn expect_close_braces(&mut self) -> Result<(), ParserError> {
        self.expect_token_with_tag(TokenTag::CloseBrace)
    }

    fn expect_semicolon(&mut self) -> Result<(), ParserError> {
        self.expect_token_with_tag(TokenTag::Semicolon)
    }

    fn expect_keyword(&mut self, expected_kwid: KeywordIdentifier) -> Result<Location, ParserError> {
        match self.get_token_with_tag(TokenTag::Keyword)? {
            Token { location, token_type: TokenType::Keyword(kwid) } => {
                if kwid == expected_kwid {
                    Ok(location)
                } else {
                    Err(ParserError::ExpectedKeyword { location, keyword_identifier: expected_kwid, actual_token: TokenTag::Keyword })
                }
            }
            Token { location, token_type } => Err(ParserError::ExpectedKeyword {
                location,
                keyword_identifier: expected_kwid,
                actual_token: token_type.tag(),
            })
        }
    }

    fn parse_identifier(&mut self) -> Result<Symbol<'a>, ParserError> {
        match self.token_provider.next() {
            Some(Ok(Token { location, token_type })) => {
                match token_type {
                    TokenType::Identifier(name) => Ok(Symbol { name, location }),
                    TokenType::Keyword(kwd) => Err(ParserError::KeywordUsedAsIdentifier { location, kwd }),
                    _ => Err(ParserError::UnexpectedToken { location, expected_token_tags: vec![TokenTag::Identifier] })
                }
            }
            Some(Err(e)) => Err(ParserError::TokenizationError(e)),
            None => Err(ParserError::UnexpectedEnd(vec![TokenTag::Identifier])),
        }
    }

    fn parse_type_expression(&mut self) -> Result<TypeExpression, ParserError> {
        let kw_loc = self.expect_keyword(KeywordIdentifier::TypeInt)?;
        Ok(TypeExpression {
            location: kw_loc,
            kind: TypeExpressionKind::Primitive(PrimitiveKind::Integer),
        })
    }

    fn parse_statement(&mut self) -> Result<Statement<'a>, ParserError> {
        self.parse_return_statement()
    }

    fn parse_return_statement(&mut self) -> Result<Statement<'a>, ParserError> {
        let kloc = self.expect_keyword(KeywordIdentifier::Return)?;
        let return_code_expr = self.parse_expression()?;
        self.expect_semicolon()?;
        Ok(Statement { location: kloc, kind: StatementKind::Return(return_code_expr) })
    }

    fn parse_expression(&mut self) -> Result<Expression<'a>, ParserError> {
        let tok = self.token_provider.peek();
        match &tok {
            Some(Ok(tok)) => {
                match tok.token_type {
                    TokenType::IntConstant(val, radix) => {
                        let loc = tok.location.clone();
                        self.token_provider.next().unwrap().expect("must be int");
                        Ok(Expression {
                            location: loc,
                            kind: ExpressionKind::IntConstant(val, radix)
                        })
                    },
                    TokenType::OpenParentheses => {
                        self.expect_open_parentheses()?;
                        let expr = self.parse_expression()?;
                        self.expect_close_parentheses()?;
                        Ok(expr)
                    },
                    TokenType::OperatorUnaryComplement => {
                        let op_loc = tok.location;
                        self.expect_token_with_tag(TokenTag::OperatorUnaryComplement)?;
                        let expr = self.parse_expression()?;
                        Ok(Expression {
                            location: op_loc,
                            kind: ExpressionKind::Unary(UnaryOperator::Complement, Box::new(expr)),
                        })
                    },
                    TokenType::OperatorMinus => {
                        let op_loc = tok.location;
                        self.expect_token_with_tag(TokenTag::OperatorMinus)?;
                        let expr = self.parse_expression()?;
                        Ok(Expression {
                            location: op_loc,
                            kind: ExpressionKind::Unary(UnaryOperator::Negate, Box::new(expr)),
                        })
                    },
                    _ => {
                        Err(ParserError::UnexpectedToken {
                            location: tok.location,
                            expected_token_tags: vec![
                                TokenTag::OpenParentheses,
                                TokenTag::OperatorUnaryComplement,
                                TokenTag::OperatorMinus,
                            ],
                        })
                    }
                }
            },
            Some(Err(e)) => Err(ParserError::TokenizationError(e.clone())),
            None => Err(ParserError::UnexpectedEnd(vec![TokenTag::IntConstant, TokenTag::OpenParentheses])),
        }
    }

    fn parse_int_constant_expression(&mut self) -> Result<Expression<'a>, ParserError> {
        let tok = self.get_token_with_tag(TokenTag::IntConstant)?;
        let tok_loc = tok.location;
        match tok.token_type {
            TokenType::IntConstant(c, rad) => Ok(Expression {
                location: tok_loc,
                kind: ExpressionKind::IntConstant(c, rad),
            }),
            _ => panic!("should not reach here"),
        }
    }

    fn expect_token_with_tag(&mut self, expected_token_tag: TokenTag) -> Result<(), ParserError> {
        self.get_token_with_tag(expected_token_tag)?;
        Ok(())
    }

    fn get_token_with_tag(&mut self, expected_token_tag: TokenTag) -> Result<Token<'a>, ParserError> {
        let token = self.token_provider.next();
        match token {
            Some(Ok(token)) => {
                let token_tag = token.token_type.tag();
                if token_tag == expected_token_tag {
                    Ok(token)
                } else {
                    Err(ParserError::UnexpectedToken { location: token.location, expected_token_tags: vec![expected_token_tag] })
                }
            }
            Some(Err(e)) => Err(ParserError::TokenizationError(e)),
            None => Err(ParserError::UnexpectedEnd(vec![expected_token_tag])),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::common::{Location, Radix};
    use crate::common::Radix::Decimal;
    use crate::lexer::Lexer;
    use crate::parser::{Expression, FunctionDefinition, Parser, ParserError, ProgramDefinition, Statement, Symbol, UnaryOperator};
    use crate::parser::ExpressionKind::{IntConstant, Unary};
    use crate::parser::StatementKind::Return;

    #[test]
    fn test_parse_program_with_tabs() {
        let src = "int	main	(	void)	{	return	0	;	}";
        let lexer = Lexer::new(src);
        let mut parser = Parser::new(lexer);
        let parsed = parser.parse();
        assert_eq!(Ok(ProgramDefinition {
            functions: vec![
                FunctionDefinition {
                    location: Location {line: 1, column: 1},
                    name: Symbol {
                        name: "main",
                        location: Location {line: 1, column: 8},
                    },
                    body: vec![
                        Statement {
                            location: Location { line: 1, column: 40 },
                            kind: Return(Expression {
                                location: Location { line: 1, column: 48 },
                                kind: IntConstant("0", Decimal),
                            }),
                        },
                    ],
                },
            ],
        }), parsed);
    }

    #[test]
    fn test_parse_multiple_functions() {
        let src = r#"
            int main(void) {
                return 2;
            }

            int foo(void) {
                return 3;
            }
        "#;
        let lexer = Lexer::new(src);
        let mut parser = Parser::new(lexer);
        let parsed = parser.parse();
        assert_eq!(Ok(ProgramDefinition {
            functions: vec![
                FunctionDefinition {
                    location: Location { line: 2, column: 13 },
                    name: Symbol { name: "main", location: Location { line: 2, column: 17 } },
                    body: vec![
                        Statement {
                            location: Location { line: 3, column: 17 },
                            kind: Return(Expression {
                                location: Location { line: 3, column: 24 },
                                kind: IntConstant("2", Decimal),
                            })
                        }
                    ],
                },
                FunctionDefinition {
                    location: Location { line: 6, column: 13 },
                    name: Symbol { name: "foo", location: Location { line: 6, column: 17 } },
                    body: vec![
                        Statement {
                            location: Location { line: 7, column: 17 },
                            kind: Return(Expression {
                                location: Location { line: 7, column: 24 },
                                kind: IntConstant("3", Decimal),
                            }),
                        }
                    ],
                }
            ],
        }), parsed)
    }

    #[test]
    fn test_parse_expressions() {
        struct ExprTestCase<'a> {
            name: &'a str,
            src: &'a str,
            expected: Result<Expression<'a>, ParserError>,
        }

        for test_case in [
            ExprTestCase {
                name: "constant base-10 integer",
                src: "100",
                expected: Ok(Expression {
                    location: Location { line: 1, column: 1 },
                    kind:  IntConstant("100", Radix::Decimal),
                }),
            },
            ExprTestCase {
                name: "complement operator",
                src: "~0xdeadbeef",
                expected: Ok(Expression {
                    location: Location { line: 1, column: 1 },
                    kind:  Unary(UnaryOperator::Complement, Box::new(Expression {
                        location: Location { line: 1, column: 2 },
                        kind: IntConstant("0xdeadbeef", Radix::Hexadecimal),
                    })),
                }),
            },
            ExprTestCase {
                name: "negation operator",
                src: "-100",
                expected: Ok(Expression {
                    location: Location { line: 1, column: 1 },
                    kind:  Unary(UnaryOperator::Negate, Box::new(Expression {
                        location: Location { line: 1, column: 2 },
                        kind: IntConstant("100", Radix::Decimal),
                    })),
                }),
            },
            ExprTestCase {
                name: "redundant parentheses around int constant",
                src: "(100)",
                expected: Ok(Expression {
                    location: Location { line: 1, column: 2 },
                    kind: IntConstant("100", Decimal),
                }),
            },
            ExprTestCase {
                name: "double complement",
                src: "~~100",
                expected: Ok(Expression {
                    location: Location { line: 1, column: 1 },
                    kind: Unary(UnaryOperator::Complement, Box::new(Expression {
                        location: Location { line: 1, column: 2 },
                        kind: Unary(UnaryOperator::Complement, Box::new(Expression {
                            location: Location { line: 1, column: 3 },
                            kind: IntConstant("100", Decimal),
                        }))
                    })),
                })
            },
            ExprTestCase {
                name: "double negation",
                src: "-(-100)",
                expected: Ok(Expression {
                    location: Location { line: 1, column: 1 },
                    kind: Unary(UnaryOperator::Negate, Box::new(Expression {
                        location: Location { line: 1, column: 3 },
                        kind: Unary(UnaryOperator::Negate, Box::new(Expression {
                            location: Location { line: 1, column: 4 },
                            kind: IntConstant("100", Decimal),
                        }))
                    })),
                })
            }
        ].into_iter() {
            let lexer = Lexer::new(test_case.src);
            let mut parser = Parser::new(lexer);
            let actual_parsed = parser.parse_expression();
            assert_eq!(test_case.expected, actual_parsed, "{}",
                format!("failed case: {name:?},\n expected: {expected:#?},\n actual: {actual:#?}",
                name = test_case.name, expected = test_case.expected, actual = actual_parsed));
        }
    }
}