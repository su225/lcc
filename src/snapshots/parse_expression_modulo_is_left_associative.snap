---
source: src/parser.rs
description: parsing expression
expression: actual.unwrap()
info: 10 % 2 % 3
---
location:
  line: 1
  column: 1
kind:
  binary:
    - modulo
    - location:
        line: 1
        column: 1
      kind:
        binary:
          - modulo
          - location:
              line: 1
              column: 1
            kind:
              int_constant:
                - "10"
                - decimal
          - location:
              line: 1
              column: 6
            kind:
              int_constant:
                - "2"
                - decimal
    - location:
        line: 1
        column: 10
      kind:
        int_constant:
          - "3"
          - decimal
